sys/kern/dev_procstat.c

```diff
@@ -0,0 +1,125 @@
+#include <sys/devfs.h>
+#include <sys/vnode.h>
+#include <sys/linker_set.h>
+#include <sys/queue.h>
+#include <sys/proc.h>
+#include <sys/thread.h>
+#include <sys/klog.h>
+#include <sys/malloc.h>
+#include <stdio.h>
+
+/* Implementation of /dev/procstat
+ *
+ * When /dev/procstat is opened it takes a snapshot of info of all existing
+ * processes.
+ *
+ * During read calls to /dev/procstat this info can be read.
+ *
+ * Example:
+ * euid   pid    ppid    pgrp    state    td_name  elfpath
+ * 0       1       0       1       R       init    /bin/ksh
+ *
+ */
+
+#define PSTAT_SIZE 1024
+#define BUF_SIZE (PSTAT_SIZE - sizeof(int))
+
+static char state[4] = {
+  [PS_NORMAL] = 'R',
+  [PS_STOPPED] = 'S',
+  [PS_DYING] = 'D',
+  [PS_ZOMBIE] = 'Z',
+};
+
+typedef struct pstat_data {
+  size_t len;
+  char buf[];
+} pstat_t;
+
+static int dev_procstat_read(file_t *f, uio_t *uio) {
+  pstat_t *pst = f->f_data;
+
+  if ((size_t)f->f_offset >= pst->len)
+    return 0;
+
+  uio->uio_offset = f->f_offset;
+  size_t len = MIN(uio->uio_resid, pst->len);
+  int error = uiomove(pst->buf, len, uio);
+
+  f->f_offset = uio->uio_offset;
+  return error;
+}
+
+static fileops_t dev_procstat_fileops = {
+  .fo_read = dev_procstat_read,
+  .fo_close = default_vnclose,
+  .fo_stat = default_vnstat,
+  .fo_ioctl = default_vnioctl,
+};
+
+static int write_to_buf(char *buf, size_t max_len, proc_t *p) {
+  if (p->p_pid == 0)
+    return 0;
+
+  return snprintf(buf, max_len, "%d\t%d\t%d\t%d\t%c\t%s\t%s\n",
+                  p->p_cred.cr_euid, p->p_pid, p->p_parent->p_pid,
+                  p->p_pgrp->pg_id, state[p->p_state], p->p_thread->td_name,
+                  p->p_elfpath);
+}
+
+static int dev_procstat_open(vnode_t *v, int mode, file_t *fp) {
+  pstat_t *pstat = kmalloc(M_TEMP, PSTAT_SIZE, 0);
+  proc_list_t *zombie_list, *proc_list;
+  proc_t *p;
+  size_t offset = 0;
+  int ret, error = EFAULT;
+
+  if (pstat == NULL)
+    return EFAULT;
+
+  get_proc_lists(&proc_list, &zombie_list);
+
+  WITH_MTX_LOCK (all_proc_mtx) {
+    TAILQ_FOREACH (p, proc_list, p_all) {
+      ret = write_to_buf(pstat->buf + offset, BUF_SIZE - offset, p);
+      if (ret < 0)
+        goto err;
+      offset += ret;
+    }
+
+    TAILQ_FOREACH (p, zombie_list, p_zombie) {
+      ret = write_to_buf(pstat->buf + offset, BUF_SIZE - offset, p);
+      if (ret < 0)
+        goto err;
+      offset += ret;
+    }
+  }
+  pstat->len = offset;
+
+  if ((error = vnode_open_generic(v, mode, fp)))
+    goto err;
+
+  fp->f_ops = &dev_procstat_fileops;
+  fp->f_data = pstat;
+  return 0;
+
+err:
+  kfree(M_TEMP, pstat);
+  return error;
+}
+
+static int dev_procstat_close(vnode_t *v, file_t *fp) {
```

Comment:
Why do we need to pass `vnode_t *v` as an argument if we won't use it?

Reply:
`vnodeops_t::v_close` is `typedef int vnode_close_t(vnode_t *v, file_t *fp);` so you need to set first argument as `vnode_t*`.